records/2024-01-22.reshaping-cicd-environments.md

```diff
@@ -0,0 +1,127 @@
+
+### Stronger responsibility for release changes on developers
+
+A lightweight changes that the process can include in the first place is to put
+more burden on the developers' shoulders on the changes approval step.
+Specifically, whatever the developer pushes to the staging environment is what
+will get released by the ops lead.
+
+This keeps the current environments as-is and does not make new automation
+additions. This is a status quo for most release di mensions except for the
+manual aspect.
+
+#### Pros
+
+- Incremental
+- Very cheap to implement
+- Lighten load on the ops lead role
+
+#### Cons
+
+- Higher risk factor of deploying to production (as it relies on multiple human interventions)
+
+### Create dedicated "Core" and "Dev" environments
+
+Currently the staging environment is used as an integration/development environment. This has been workng, but is not aligned with best practices, as it leaves staging in a non-production like state. With the introduction of the new "Dev" environment, it became possible to add an additional environment into the process that allows staging to always remain in the same state as production. 
+
+The new "Dev" envorinment as it exists today is not suitable for such a role because it is constantly being created/destroyed/altered by the Core team in order to test and develop infrastructure changes. It would not be reliable enough for the App team to develop against. 
+
+An additional environment could be created that is reserved for the App team's development and integration, and the existing "Dev" environment could be reclassified as the "Core" environment to identify it as being used by the Core team for infrastructure development.
+
+![Potential Core and Dev Environment Solution](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/2024-01-22.reshaping-cicd-environments/core-dev-env.png?raw=true "Potential Core and Dev Environment Solution")
+
+#### Pros
+  - Staging could be used to test the actual production release package and the associated process, which is currently not being tested.
+  - The release process to staging could be automated completely. This would provide a safety net for automating releases to production. I.e: at end of day, automatically create a release package from Dev, and automatically deploy this package to staging. Run automated testing overnight. If the tests pass, there is a high degree of certainty that it will release successfully in production, and thus does not require intervention from the team/Ops Lead.
+  - Application development would not be hindered by the Core team's infrastructure changes
+
+#### Cons
+  - Potentially increased cloud costs, but there are ways to reduce/eliminate this
+  - Increased complexity for Core team with environment management
+  - Increased complexity for release process
+
```

Comment:
- We could in theory tag the release and run performance tests overnight on staging (I think we already do this?). And then release to prd in the morning? vs adding an env overhead.

Reply:
It's a possibility that I will include as a different option, probably under a release train. Each will have their own set of pros and cons. What would get picked depends on what the team feels comfortable with in terms of risk, complexity, effort, budget, process alignment and automation (if I can summarize the initial goals that way).